### PR TITLE
[issue-43] implement a support-scheduler-client

### PR DESCRIPTION
**Pull request by yanghua**
Signed-off-by: yanghua <yanghua1127@gmail.com>

### CODE REVIEW

**Comment by yanghua** [comment]
hi @jpwhitemn , I just try to write a support-scheduler go client. It contains : 

* a `client.go` file : to send schedule and schedule event http request to scheduler microservice
* a `client_test.go` file : to mock a http serverand test the two function

Would you please review this PR and give some suggestion?

File: support/scheduler-client/client.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*******************************************************************************
+ * Copyright 2017 Dell Inc.
```
Comment (drasko) [inline support/scheduler-client/client.go]: You are not from Dell, if you are not extending other people work, put your copyright info.

File: support/scheduler-client/client.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	RemoteScheduleEventUrl string
+	RemoteCallbackAlertUrl string
+	OwningService          string
+}
+
+// Function to send a schedule to the remote scheduler server
+func (schedulerClient SchedulerClient) SendSchedule(schedule models.Schedule) error {
+	client := &http.Client{}
+
+	// get the json request body
```
Comment (drasko) [inline support/scheduler-client/client.go]: Drop unnecessary comment

File: support/scheduler-client/client.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	return doPost(schedulerClient.RemoteScheduleUrl, bytes.NewBuffer(requestBody), client)
+}
+
+// Function to send a schedule event to the remote scheduler server
+func (schedulerClient SchedulerClient) SendScheduleEvent(scheduleEvent models.ScheduleEvent) error {
+	client := &http.Client{}
+
+	// get the json request body
```
Comment (drasko) [inline support/scheduler-client/client.go]: Unnecessary comment

File: support/scheduler-client/client.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		fmt.Println(err.Error())
+		return err
+	}
+
+	return doPost(schedulerClient.RemoteScheduleEventUrl, bytes.NewBuffer(requestBody), client)
+}
+
+// Function to do post request
+func doPost(url string, binaryReqBody io.Reader, client *http.Client) error {
+	// create request
```
Comment (drasko) [inline support/scheduler-client/client.go]: Drop evident comments

File: support/scheduler-client/client.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	return nil
+}
+
+// Function to actually make the HTTP request
+func sendRequest(client *http.Client, req *http.Request) {
+	resp, err := client.Do(req)
+	if err == nil {
+		defer resp.Body.Close()
+		resp.Close = true
+	} else {
```
Comment (drasko) [inline support/scheduler-client/client.go]: If err return and drop `else`

File: support/scheduler-client/client.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		fmt.Println(err.Error())
```
Comment (drasko) [inline support/scheduler-client/client.go]: No printing on the command line - use logs.

Comment (yanghua) [reply support/scheduler-client/client.go]: @drasko thanks for reviewing. Does you mean use edgex's `LoggingClient`?

**Comment by yanghua** [comment]
Hi @drasko I have do some change base on your suggestion, Please review again. Thanks!

File: support/scheduler-client/client_test.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
+/*******************************************************************************
+ * Copyright 2017 Dell Inc.
```
Comment (drasko) [inline support/scheduler-client/client_test.go]: It's not Dell

File: support/scheduler-client/client.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
+ * Copyright 2017 Tencent Inc.
```
Comment (drasko) [inline support/scheduler-client/client.go]: Consider using SPDX-style comments (https://spdx.org/) to improve readability and reduce comment bloating: https://github.com/edgexfoundry/edgex-go/blob/master/export/client/server.go#L1-L5

**Comment by yanghua** [comment]
cc @drasko I have changed the license style, please review, Thanks!

File: support/scheduler-client/client_test.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+			t.Errorf(TestUnexpectedMsgFormatStrForBoolVal, receivedSchedule.RunOnce, TestScheduleRunOnce)
+		}
+
+	}))
+
+	defer ts.Close()
+	uriPrefix := ts.URL
+
+	scheduleClient := SchedulerClient{
+		RemoteScheduleUrl: uriPrefix + "/api/v1/schedule",
```
Comment (drasko) [inline support/scheduler-client/client_test.go]: Use defined consts for strings (you already defined `TestExpectedScheduleApiPath`)

File: support/scheduler-client/client_test.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+			t.Errorf(TestUnexpectedMsgFormatStr, receivedScheduleEvent.Addressable.Protocol, TestScheduleEventAddressableProtocol)
+		}
+
+	}))
+
+	defer ts.Close()
+	uriPrefix := ts.URL
+
+	scheduleClient := SchedulerClient{
+		RemoteScheduleEventUrl: uriPrefix + "/api/v1/scheduleevent",
```
Comment (drasko) [inline support/scheduler-client/client_test.go]: Use defined consts for strings.

**Comment by yanghua** [comment]
cc @drasko

File: support/scheduler-client/client.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+// Common http const
+const (
+	HttpPostMethod     = "POST"
+	ContentType        = "Content-Type"
+	ContentTypeJsonVal = "application/json"
+)
+
+var loggingClient logger.LoggingClient = logger.NewClient(SchedulerClientName, false, "")
```
Comment (feclare) [inline support/scheduler-client/client.go]: Libraries should try to not write logs directly. In this case you are only logging before returning an error, so the user of this client will have the same information to create the log.
SchedulerClientName should not be needed.

File: support/scheduler-client/client.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+// Struct to represent the scheduler client
+type SchedulerClient struct {
+	RemoteScheduleUrl      string
+	RemoteScheduleEventUrl string
```
Comment (feclare) [inline support/scheduler-client/client.go]: The only needed configuration should be the hostname and port, not urls. The client should know how to create the complete url from the hostname and port.

Comment (yanghua) [reply support/scheduler-client/client.go]: @feclare I agree this opinion, I just followed the notification's client, I will change this.

File: support/scheduler-client/client.go (hunk 11 above)
Comment (yanghua) [reply support/scheduler-client/client.go]: @feclare  you mean I don't need to use `LoggingClient` in scheduler's client?

**Comment by yanghua** [comment]
@feclare I will add the update and remove function and test case, soon.

File: support/scheduler-client/client.go (hunk 11 above)
Comment (feclare) [reply support/scheduler-client/client.go]: Yes, it is preferable to not log directly from the libs if not really needed. Like in this case.

**Comment by yanghua** [comment]
@feclare please review again, thanks!

File: support/scheduler-client/client.go [hunk 13]
```diff
@@ -8,31 +8,33 @@ package scheduler
...
+)
 
 // Struct to represent the scheduler client
 type SchedulerClient struct {
-	RemoteScheduleUrl      string
-	RemoteScheduleEventUrl string
-	RemoteCallbackAlertUrl string
-	OwningService          string
+	SchedulerServiceHost string
+	SchedulerServicePort string
```
Comment (feclare) [inline support/scheduler-client/client.go]: Change SchedulerService port to int

File: support/scheduler-client/client.go [hunk 14]
```diff
@@ -41,11 +43,51 @@ func (schedulerClient SchedulerClient) SendSchedule(schedule models.Schedule) er
 
 	requestBody, err := schedule.MarshalJSON()
 	if err != nil {
-		loggingClient.Error(err.Error())
 		return err
 	}
 
-	return doPost(schedulerClient.RemoteScheduleUrl, bytes.NewBuffer(requestBody), client)
+	remoteScheduleUrl := ""
+	if schedulerClient.SchedulerServicePort != "" {
```
Comment (feclare) [inline support/scheduler-client/client.go]: Port is always present. No need for two different cases.

File: support/scheduler-client/client_test.go [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+// Test common const
+const (
+	TestUnexpectedMsg                    = "unexpected result"
+	TestUnexpectedMsgFormatStr           = "unexpected result, active: '%s' but expected: '%s'"
+	TestUnexpectedMsgFormatStrForBoolVal = "unexpected result, active: '%t' but expected: '%t'"
+)
+
+// Test comm api path
+const (
+	TestExpectedScheduleApiPath      = "/api/v1/schedule"
```
Comment (feclare) [inline support/scheduler-client/client_test.go]: The same value as ScheduleApiPath and ScheduleEventApiPath, remove TestExpectedScheduleApiPath and TestExpectedScheduleEventApiPath

File: support/scheduler-client/client_test.go [hunk 16]
```diff
@@ -0,0 +1,425 @@
...
+	TestScheduleEventParameters          = ""
+	TestScheduleEventService             = "notifications"
+	TestScheduleEventSchedule            = "testSchedule"
+	TestScheduleEventAddressableName     = "MQTT"
+	TestScheduleEventAddressableProtocol = "MQTT"
+	TestScheduleEventIdForTest           = "testScheduleEventId"
+)
+
+// Test method : SendSchedule
+func TestSchedulerClient_SendSchedule(t *testing.T) {
```
Comment (feclare) [inline support/scheduler-client/client_test.go]: Rename function tests from TestSchedulerClient_SendSchedule to something like TestSendSchedule

File: support/scheduler-client/client_test.go [hunk 17]
```diff
@@ -0,0 +1,425 @@
...
+	h := strings.Split(u.Host, ":")
+
+	scheduleClient := SchedulerClient{
+		SchedulerServiceHost: h[0],
+		SchedulerServicePort: h[1],
+		OwningService:        "notifications",
+	}
+
+	error := scheduleClient.RemoveScheduleEvent(TestScheduleEventIdForTest)
+	println(error)
```
Comment (feclare) [inline support/scheduler-client/client_test.go]: Remove println commands and if error is not nil report error.

**Comment by yanghua** [comment]
cc @feclare

**Comment by feclare** [comment]
Thanks @yanghua for all the code changes.

**Pull request merged**